app: Enable Memfault by default (#367)
nrfconnect/ncs-serial-modem: doc/app/at_nrfcloud.rst

------------

The test command is not supported.

.. _SM_AT_NRFCLOUDOBS:

nRF Cloud observability
=======================

The ``#XNRFCLOUDOBS`` command controls the Memfault observability data that the device collects (metrics, events, logs, and, in builds that include it, a coredump) and its upload to nRF Cloud over the CoAP transport.

.. note::
   To use ``#XNRFCLOUDOBS``, the following preconditions apply:

   * The :ref:`CONFIG_SM_NRF_CLOUD_OBSERVABILITY <CONFIG_SM_NRF_CLOUD_OBSERVABILITY>` Kconfig option must be enabled.
     It depends on :kconfig:option:`CONFIG_MEMFAULT_USE_NRF_CLOUD_COAP`.
   * The operations that access the network (``3`` and ``4``) require a connection to nRF Cloud.
     See ``AT#XNRFCLOUD``.

   Upload is host-driven.
   Set :kconfig:option:`CONFIG_MEMFAULT_PERIODIC_UPLOAD_ENABLED_DEFAULT` to ``n`` to have the periodic upload start disabled, and enable it at runtime with the operation ``1``.

Set command
-----------

The set command controls the observability data collection and upload.

Syntax
~~~~~~

::

>    AT#XNRFCLOUDOBS=<op>[,<base64_chunk>[,<project_key>]]

Reviewer: A month ago I didn't have much opinion on how to structure the AT command(s) but we need to "design" them in a manner that we can extend them as conveniently as possible. Since `<base64_chunk>` and `<project_key>` are used only with certain `<op>`. it makes backwards compatible updates a bit clumsy. E.g., how would you add periodic upload interval parameter to this command in the future?
I think we should consider splitting this a bit.

------------

The test command is not supported.

.. _SM_AT_NRFCLOUDOBS:

nRF Cloud observability
=======================

The ``#XNRFCLOUDOBS*`` commands control the Memfault data that the device collects (metrics, events, logs, and, in builds that include it, a coredump) and its upload to nRF Cloud over the CoAP transport.

.. note::
   To use the ``#XNRFCLOUDOBS*`` commands, the following preconditions apply:

>    * The :ref:`CONFIG_SM_NRF_CLOUD_OBSERVABILITY <CONFIG_SM_NRF_CLOUD_OBSERVABILITY>` Kconfig option must be enabled.

Reviewer 2: Isn't this is on by default? If so, no need to mention it here.

>      It depends on :kconfig:option:`CONFIG_MEMFAULT_USE_NRF_CLOUD_COAP`.

Reviewer 2: ```suggestion
```

------------

The test command is not supported.

.. _SM_AT_NRFCLOUDOBS:

nRF Cloud observability
=======================

The ``#XNRFCLOUDOBS*`` commands control the Memfault data that the device collects (metrics, events, logs, and, in builds that include it, a coredump) and its upload to nRF Cloud over the CoAP transport.

.. note::
   To use the ``#XNRFCLOUDOBS*`` commands, the following preconditions apply:

   * The commands that access the network (``#XNRFCLOUDOBSUPLOAD`` and ``#XNRFCLOUDOBSFORWARD``) require a connection to nRF Cloud.
     See ``AT#XNRFCLOUD``.

>    * ``#XNRFCLOUDOBSDEVINFO``, ``#XNRFCLOUDOBSCRASH`` and ``#XNRFCLOUDOBSEXPORT`` additionally require the :ref:`CONFIG_SM_NRF_CLOUD_OBSERVABILITY_DEBUG <CONFIG_SM_NRF_CLOUD_OBSERVABILITY_DEBUG>` Kconfig option, which is disabled by default.

Reviewer 3: ```suggestion
   * ``#XNRFCLOUDOBSDEVINFO``, ``#XNRFCLOUDOBSCRASH``, and ``#XNRFCLOUDOBSEXPORT`` additionally require the :ref:`CONFIG_SM_NRF_CLOUD_OBSERVABILITY_DEBUG <CONFIG_SM_NRF_CLOUD_OBSERVABILITY_DEBUG>` Kconfig option, which is disabled by default.
```

   Upload is host-driven.

>    The automatic upload starts disabled, and the host enables it with ``AT#XNRFCLOUDOBSAUTO=1`` or uploads on demand with ``AT#XNRFCLOUDOBSUPLOAD``.

Reviewer 3: ```suggestion
   Automatic uploads are disabled by default.
   The host enables it with ``AT#XNRFCLOUDOBSAUTO=1`` or initiate uploads on demand with ``AT#XNRFCLOUDOBSUPLOAD``.
```

> The ``<project_key>`` parameter, which several of the commands accept, is a string.

Reviewer 3: ```suggestion
The ``<project_key>`` parameter, used by several commands, is a string.
```

The ``#XNRFCLOUDOBSAUTO`` command configures the automatic upload of the buffered observability data.

The configuration is persistent, so it survives a reboot.

Set command
-----------

The set command enables or disables the automatic upload and configures its interval.

Syntax
~~~~~~

::

   AT#XNRFCLOUDOBSAUTO=<enable>[,<interval_seconds>][,<project_key>]

* The ``<enable>`` parameter can have the following integer values:

  * ``0`` - Disable the automatic upload.
  * ``1`` - Enable the automatic upload.

* The ``<interval_seconds>`` parameter is an integer from ``60`` to ``86400``.
  It is the interval between two uploads.
  When it is omitted, the stored interval is kept.
  Its initial value is set by the :ref:`CONFIG_SM_NRF_CLOUD_OBSERVABILITY_AUTO_INTERVAL_SECONDS <CONFIG_SM_NRF_CLOUD_OBSERVABILITY_AUTO_INTERVAL_SECONDS>` Kconfig option.

* The ``<project_key>`` parameter is a string.
  When it is omitted, the stored project key is kept, and an empty string clears it.

The first upload runs when the interval expires, not when the automatic upload is enabled.
An upload that falls while there is no connection to nRF Cloud is skipped, and the next one is scheduled as usual.

> The automatic upload is silent: it sends no unsolicited notification.

Reviewer 3: ```suggestion
The automatic upload is silent and does not send any unsolicited notifications..
```

Unsolicited notification
~~~~~~~~~~~~~~~~~~~~~~~~

::

   #XNRFCLOUDOBSUPLOAD: <result>[,<bytes>]

* The ``<result>`` parameter is an integer.

  * ``0`` - Success.
    The ``<bytes>`` parameter follows and indicates the number of bytes uploaded, which is ``0`` when there was nothing buffered.
  * ``-1`` - Failure.
    The error code is shown in the log.

Example
~~~~~~~

::

  AT#XNRFCLOUDOBSUPLOAD

  OK

  #XNRFCLOUDOBSUPLOAD: 0,108

Read command
------------

The read command is not supported.

Test command
------------

The test command returns the supported syntax.

Syntax
~~~~~~

::

   AT#XNRFCLOUDOBSUPLOAD=?

Response
~~~~~~~~

::

   #XNRFCLOUDOBSUPLOAD: <project_key>

.. _SM_AT_NRFCLOUDOBSHEARTBEAT:

Metrics heartbeat #XNRFCLOUDOBSHEARTBEAT
========================================

The ``#XNRFCLOUDOBSHEARTBEAT`` command collects and finalizes a metrics heartbeat.

The data is buffered on the device until it is uploaded, so the command does not require a connection to nRF Cloud.

The heartbeat carries the LTE metrics that the modem reports: the modem firmware version, the network operator, RSRP, SNR, the current band, and the transmitted and received kilobytes.

> Values that the modem does not report, for example when it is deactivated, are left out of the heartbeat.

Reviewer 3: ```suggestion
The heartbeat carries the LTE metrics that the modem reports, such as the modem firmware version, the network operator, RSRP, SNR, the current band, and the transmitted and received data in kilobytes.
Values not reported by the modem, such as when it is deactivated, are excluded from the heartbeat.
```

  OK

Read command
------------

The read command is not supported.

Test command
------------

The test command is not supported.

.. _SM_AT_NRFCLOUDOBSCRASH:

Forced crash #XNRFCLOUDOBSCRASH
===============================

The ``#XNRFCLOUDOBSCRASH`` command forces a crash, so that the coredump capture and upload can be tested.

This command requires the :ref:`CONFIG_SM_NRF_CLOUD_OBSERVABILITY_DEBUG <CONFIG_SM_NRF_CLOUD_OBSERVABILITY_DEBUG>` Kconfig option.

Set command
-----------

The set command crashes the application.

Syntax
~~~~~~

::

   AT#XNRFCLOUDOBSCRASH[=<type>]

* The ``<type>`` parameter is an integer from ``0`` to ``4``.
  It defaults to ``0``, an assertion failure.

The device crashes and no response is returned, unless the crash type is invalid.

Read command
------------

The read command is not supported.

Test command
------------

The test command returns the supported syntax.

Syntax
~~~~~~

::

   AT#XNRFCLOUDOBSCRASH=?

Response
~~~~~~~~

::

   #XNRFCLOUDOBSCRASH: <type>

.. _SM_AT_NRFCLOUDOBSEXPORT:

Chunk export #XNRFCLOUDOBSEXPORT
================================

> The ``#XNRFCLOUDOBSEXPORT`` command prints the buffered Memfault chunks to the AT interface instead of uploading them, in the Memfault chunk export format, which the Memfault tooling can parse.

Reviewer 3: ```suggestion
The ``#XNRFCLOUDOBSEXPORT`` command prints the buffered Memfault chunks to the AT interface instead of uploading them in the Memfault chunk export format, which the Memfault tooling can parse.
```

------------

The test command is not supported.

.. _SM_AT_NRFCLOUDOBS:

nRF Cloud observability
=======================

The ``#XNRFCLOUDOBS*`` commands control the Memfault observability data that the device collects (metrics, events, logs, and, in builds that include it, a coredump) and its upload to nRF Cloud over the CoAP transport.

.. note::
   To use the ``#XNRFCLOUDOBS*`` commands, the following preconditions apply:

   * The :ref:`CONFIG_SM_NRF_CLOUD_OBSERVABILITY <CONFIG_SM_NRF_CLOUD_OBSERVABILITY>` Kconfig option must be enabled.
     It depends on :kconfig:option:`CONFIG_MEMFAULT_USE_NRF_CLOUD_COAP`.
   * The commands that access the network (``#XNRFCLOUDOBSUPLOAD`` and ``#XNRFCLOUDOBSFORWARD``) require a connection to nRF Cloud.
     See ``AT#XNRFCLOUD``.
   * ``#XNRFCLOUDOBSDEVINFO``, ``#XNRFCLOUDOBSCRASH`` and ``#XNRFCLOUDOBSEXPORT`` additionally require the :ref:`CONFIG_SM_NRF_CLOUD_OBSERVABILITY_DEBUG <CONFIG_SM_NRF_CLOUD_OBSERVABILITY_DEBUG>` Kconfig option, which is disabled by default.

   Upload is host-driven.
   The automatic upload starts disabled, and the host enables it with ``AT#XNRFCLOUDOBSAUTO=1`` or uploads on demand with ``AT#XNRFCLOUDOBSUPLOAD``.

> The ``<project_key>`` parameter, which several of the commands accept, is a string.

Author: @noahp are you ok with this explanation on the project key usage? Should we link this to any Memfault's document?

Reviewer 4: You can link here: https://docs.memfault.com/docs/platform/data-routes

and here: ("Find your project key in Project Settings: https://app.memfault.com/organizations/-/projects/-/settings/data-routes")

otherwise, lgtm!

Author: Added!

Example
~~~~~~~

::

  AT#XNRFCLOUDOBSAUTO=1,600

  OK
  AT#XNRFCLOUDOBSAUTO?

  #XNRFCLOUDOBSAUTO: 1,600,""

  OK

Test command
------------

The test command returns the supported syntax.

Syntax
~~~~~~

::

   AT#XNRFCLOUDOBSAUTO=?

Response
~~~~~~~~

::

   #XNRFCLOUDOBSAUTO: (0,1),(60-86400),<project_key>

.. _SM_AT_NRFCLOUDOBSUPLOAD:

On-demand upload #XNRFCLOUDOBSUPLOAD
====================================

The ``#XNRFCLOUDOBSUPLOAD`` command uploads the buffered observability data to nRF Cloud.

The captured logs are collected before the upload, so that they are included.

Set command
-----------

The set command uploads the buffered data.

Syntax
~~~~~~

::

   AT#XNRFCLOUDOBSUPLOAD[=<project_key>]

The command returns ``OK`` immediately and the upload runs asynchronously.
When it completes, an unsolicited notification is sent.

Unsolicited notification
~~~~~~~~~~~~~~~~~~~~~~~~

::

   #XNRFCLOUDOBSUPLOAD: <result>[,<bytes>]

* The ``<result>`` parameter is an integer.

  * ``0`` - Success.
    The ``<bytes>`` parameter follows and indicates the number of bytes uploaded, which is ``0`` when there was nothing buffered.

>   * A negative value - A negative error code.

Reviewer: We've used -1 most of the time as an error value (see e.g. #XSENDNTF). Otherwise we basically need to list possible error codes. The error value usually doesn't matter for the host. And developer is looking at the log anyway where the code should be shown.

Author: Let me change them then!

------------

The read command is not supported.

Test command
------------

The test command returns the supported syntax.

Syntax
~~~~~~

::

   AT#XNRFCLOUDOBSUPLOAD=?

Response
~~~~~~~~

::

   #XNRFCLOUDOBSUPLOAD: <project_key>

.. _SM_AT_NRFCLOUDOBSHEARTBEAT:

Metrics heartbeat #XNRFCLOUDOBSHEARTBEAT
========================================

The ``#XNRFCLOUDOBSHEARTBEAT`` command collects and finalizes a metrics heartbeat.

The data is buffered on the device until it is uploaded, so the command does not require a connection to nRF Cloud.

Set command
-----------

The set command collects a heartbeat.

Syntax
~~~~~~

::

   AT#XNRFCLOUDOBSHEARTBEAT

Example
~~~~~~~

::

  AT#XNRFCLOUDOBSHEARTBEAT

  OK

Read command
------------

The read command is not supported.

Test command
------------

The test command returns ``OK``.

Syntax
~~~~~~

::

   AT#XNRFCLOUDOBSHEARTBEAT=?

.. _SM_AT_NRFCLOUDOBSFORWARD:

Chunk forwarding #XNRFCLOUDOBSFORWARD
=====================================

The ``#XNRFCLOUDOBSFORWARD`` command forwards a Memfault chunk produced by the host to nRF Cloud, using the |SM| connection.

.. note::
   nRF Cloud attributes the chunk to the authenticated device.

>    The chunk itself carries the identity of the device that produced it, which is what Memfault uses to attribute the data.

Author: @noahp should I extend the explanation that the host should enable MEMFAULT_EVENT_INCLUDE_DEVICE_SERIAL?

Reviewer 4: that's only necessary if they want to override the default Device Serial provided by the Serial Modem, and use a different Device Serial for the host observability data- so I'd mention that in the description, so it's clear what they need to do if they want that 👍

Author: Done!

------------

The read command is not supported.

Test command
------------

The test command returns the supported syntax.

Syntax
~~~~~~

::

   AT#XNRFCLOUDOBSCRASH=?

Response
~~~~~~~~

::

   #XNRFCLOUDOBSCRASH: <type>

.. _SM_AT_NRFCLOUDOBSEXPORT:

Chunk export #XNRFCLOUDOBSEXPORT
================================

The ``#XNRFCLOUDOBSEXPORT`` command prints the buffered Memfault chunks to the AT interface instead of uploading them, in the Memfault chunk export format, which the Memfault tooling can parse.

This command requires the :ref:`CONFIG_SM_NRF_CLOUD_OBSERVABILITY_DEBUG <CONFIG_SM_NRF_CLOUD_OBSERVABILITY_DEBUG>` Kconfig option.

.. note::
   The command consumes the chunks.
   They are no longer available for ``#XNRFCLOUDOBSUPLOAD`` and the automatic upload.

Set command
-----------

The set command prints the buffered chunks.

Syntax
~~~~~~

::

   AT#XNRFCLOUDOBSEXPORT

Response
~~~~~~~~

The buffered chunks are returned, one per line::

   MC:<base64_chunk>:

Example
~~~~~~~

::

  AT#XNRFCLOUDOBSHEARTBEAT

  OK
  AT#XNRFCLOUDOBSEXPORT

  MC:CAKnAgIDAQpqdGVzdHNlcmlhbA==:

  OK

Read command
------------

The read command is not supported.

Test command
------------

> The test command returns ``OK``.

Reviewer: The test commands that are not returning anything should just not be supported like elsewhere.
